### ISSUE TITLE
Fixtures do not support JavaScript in .js files

### ISSUE THREAD

**Issue by dkreft** [cypress-io/cypress#1271]
- Operating System:

Darwin Kernel Version 15.6.0: Mon Nov 13 21:58:35 PST 2017; root:xnu-3248.72.11~1/RELEASE_X86_64

- Cypress Version:

1.4.2

- Browser Version:

Electron 53, Chrome 64

### Is this a Feature or Bug?

Bug.

### Current behavior:

Any attempt to use valid ES5 (e.g. `var`, `require`) let alone ES6 (`import`, `export`, `const`) results in syntax errors, e.g.:

```
Error: 'admin.js' is not a valid JavaScript object.SyntaxError: Unexpected token var

Because this error occurred during a 'before each' hook we are skipping the remaining tests in the current suite: 'when login is valid'
```

The only "javascripty" construct that seems to be allowed is `module.exports =`.

### Desired behavior:

`.js` fixtures should support valid JS syntax (ES6, preferably).

#### Create a fixture containing valid ES6 syntax

```js
const user = 'Joe'

export default {
  user,
}
```

Then use it:

```js
  cy.fixture('auth.js')
```

Result:

```
Error: 'auth.js' is not a valid JavaScript object.
auth.js:3
export default {
^
ParseError: Unexpected token
```

#### Create a fixture using valid ES5 code

```js
var user = 'Joe';

module.exports = {
  user: user
};
```

Result:

```
Error: 'auth.js' is not a valid JavaScript object.SyntaxError: Unexpected token var
```

**Comment by tomatau** [comment]
please fix this

**Comment by bensampaio** [comment]
any news on this?

**Comment by zebulonj** [comment]
It'd be lovely if `*.js` fixtures were simply executed like any normal JavaScript, and they were expected to export the same fixture data that might otherwise be JSON encoded in a `*.json` fixture. Prime use case: using [`faker`](https://github.com/marak/Faker.js/) in a `.js` fixture to generate data.

**Comment by sandro-pasquali** [comment]
@brian-mann Sincere question: why are `.js` fixtures *not* treated as a typical `.js` files? Is there something about the context in a `cy` test which precludes that? Just trying to understand, and would love to help fix. Pretty normal (for me at least) to dynamically generate fixtures at runtime via `.js`. Kinda need this.

**Comment by brian-mann** [comment]
@sandro-pasquali the reason is that there are two contexts - the browser context and node context.

Because of this, it's not as simple as what you'd expect. Fixtures are utilized inside of the browser, which means you use them like this... `cy.fixture('foo.js', arg1, arg2, arg3)` right?

That's fine except... how is the `foo.js`evaluated? If it's evaluated in the browser context then that fixture file needs to **already be part** of the bundle that is served along with spec files. That way it could go through the normal spec preprocessor rules so you could use your typescript, webpack modules, etc on it. But wait: how does it become part of the browser bundle if you don't ever `require(...)` or `import` it directly in your spec files?

Alternatively, we could evaluate `foo.js` in the `node` context. That's possible, but then we run afoul of the same preprocessor rules. It would mean your `specs` are evaluated with preprocessors but then fixtures aren't. We might could add fixtures to those rules too which would would enable you to use typescript on there, but it really doesn't make sense in cases like with webpack because  this `foo.js` isn't actually being served to the browser.

Another problem with the node context approach is that arguments would have to be JSON serializable. That's probably *not* that big of a problem but it means you cannot pass functions or instances or even `undefined` across the wire.

I'm probably more okay with opting to go with the `node` context approach as opposed to the browser.

There has been some dialog amongst our team to effectively rewrite the entire fixtures approach in Cypress because we feel it's been large superseded by better approaches and initially was sort of a "stop-gap" feature invented years ago.

**Comment by brian-mann** [comment]
@sandro-pasquali forgot to mention that you can interop between the browser + node context flawlessly with `cy.task`. You could effectively design or come up with your own fixture implementation without using `cy.fixture` at all. Just defer to a `cy.task`, pass in arguments, and then dynamically call into whatever fixture or node.js file you want to do.

This is effectively the approach I described above as the 2nd option. The only downside is that you'll have to write your node code for node `8.2.1` and it will not support `typescript` at all.

**Comment by simonhaenisch** [comment]
I'm running into similar issues with something like this:

```js
[
  { /* ... */ },
];
```

> Error: 'notifications' is not a valid JavaScript object. SyntaxError: Unexpected token `;`

Removing the trailing semicolon fixes it, however my prettier editor plugin auto-appends it on save.

I was expecting the fixtures to become part of the browser bundle and also tried `export default []` first. I see the issue with having to import them somewhere, however i'm wondering what the actual issue is with importing them?

This is more JS-like any (passing the fixture as an object reference):

```js
import notificationsFixture from '../fixtures/notifications';

describe('Notifications', () => {
  it('lists notifications', () => {
    cy.server();
    cy.route('GET', '/api/notifications', notificationsFixture);
    
    // or
    cy.fixture(notificationsFixture).as('notifications');
    cy.route('GET', '/api/notifications', '@notifications');

    // ...
  });
});
```

This could even become a backwards-compatible feature (i. e., check whether the fixture is a string or an object)?

**Comment by chochihim** [comment]
Maybe we need to update the document to state that js fixture is not what people normally expect.
I am quite surprised that
this fixture works:
```js
module.exports = {
  a: 1
}
```
but not this (with semicolon automatically added by my editor setting)
```js
module.exports = {
  a: 1
};
```
This throws `Error: 'myFixture' is not a valid JavaScript object.SyntaxError: Unexpected token ;`):

**Comment by SLOBYYYY** [comment]
Any news about this? It's really hard to maintain our fixtures since they are quite big with only slight changes here and there. One change in any of them and you have to update all of them.

**Comment by bahmutov** [comment]
Take a look at cy.task - it can be used to generate any object in node context, including fixture json files that you could load later. Or use cy.writeFile to do the same.  Or even JavaScript right from the tests and instead of fixture stub in cy.route use the created object.

**Comment by younesea** [comment]
Thanks @bahmutov for mentioning cy.task.

I had the same issue as @SLOBYYYY and created a task that fit my needs: 


**plugins/index.js**
```
module.exports = (on, config) => {
  on('task', {
    fixture (fixtures) {
      const result = {}

      if (Array.isArray(fixtures)) {
        fixtures.forEach(fixture => {
          result[fixture] = require('../fixtures/elements/' + fixture + '.js')
        })
      } else {
        result[fixtures] = require('../fixtures/elements/' + fixtures + '.js')
      }

      return result
    }
  })

  return config
}
```

**So in your tests you can use it like this**

```
// Getting one fixture
cy.task('fixture', 'user').then(v => {
      console.log(v) // outputs {user: {}}
})

// Getting multiple fixtures
cy.task('fixture', ['user', 'product']).then(v => {
      console.log(v) // outputs { user: {}, product: {]}
})
```

You can adjust the task so that it works for your project.

**Comment by OneHatRepo** [comment]
I could not get the task option to work, as my .js fixture files had ES6 imports that were consantly failing. In the end, I just did a standard import of each fixture file at the top of each spec file and then cloned the data in a beforeEach.

```
import props from '../../../fixtures/Models/Helpers/props';
import _ from 'lodash';

....

	beforeEach(function() {
		cy.then(function() {
			return _.cloneDeep(props);
		}).as('props');
	});
```
The fixture was then available in each test as **this.props**.

**Comment by renatoruk** [comment]
> > Maybe we need to update the document to state that js fixture is not what people normally expect.
> > I am quite surprised that
> > this fixture works:
> > ```js
> > module.exports = {
> >   a: 1
> > }
> > ```
> > 
> > 
> > but not this (with semicolon automatically added by my editor setting)
> > ```js
> > module.exports = {
> >   a: 1
> > };
> > ```
> > 
> > 
> > This throws `Error: 'myFixture' is not a valid JavaScript object.SyntaxError: Unexpected token ;`
> 
> Took me an hour to find :|

Yep, same here. I actually resorted to having a `__fixtures__` directory inside `integration/{moduleName}` directory. What is the use case for the fixtures you need? I am trying to mock some responses, so for now it seems better to use them directly as a value to cy.route options parameter.
```js
import {response} from "__fixtures__/my-response.js";

cy.route({
  method: "GET", // Route all GET requests
  url: "https://api.my-app.com/**",
  response: response.body,
});
``` 

But, I am not really sure will this break something in the future, I haven't researched it fully yet. Does anyone use response mocks like this?

**Comment by JenChampagne** [comment]
Given that fixtures are meant to be data we store externally to tests for the purposes of use inside tests, I can't think of any situations where I would want a `js` fixture to be executed in the browser context. If I require that type of execution, I would place it in a command where I know it has access to the browser/DOM.

There absolutely may be use cases which make more sense to have them executed in a browser context that I am not aware of, but refusing to do either doesn't seem like a good option.

It would be nice if it were at least possible to set which context is used in the cypress config file. Maybe even have an export value that specifies the context.
```
export default {
  some: 'data',
  x: fakedX(),
  cypress: {
    moduleContext: 'node',
  },
};
```

For now, I am not using fixtures at all and doing similar to @renatoruk in that I'm using `import` or `require` directly in individual tests as needed.

**Comment by JenChampagne** [comment]
Given that fixtures are meant to be data we store externally to tests for the purposes of use inside tests, I can't think of any situations where I would want a `js` fixture to be executed in the browser context. If I require that type of execution, I would place it in a command where I know it has access to the browser/DOM.

There absolutely may be use cases which make more sense to have them executed in a browser context that I am not aware of, but refusing to do either doesn't seem like a good option.

It would be nice if it were at least possible to set which context is used in the cypress config file.

For now, I am not using fixtures at all and doing similar to @renatoruk in that I'm using import or require directly in individual tests as needed.

**Comment by mesqueeb** [comment]
What's the recommended way to import mock data into tests which rendered by triggering a function?
Eg. `mockUser()` or `new Array(100).fill(null).map(mockUser)`

Could I just create a TS file that exports these functions that create user mocks and import and execute the function inside cypress tests?

**Comment by bahmutov** [comment]
Yes you can import or require these functions and call them to generate mock data

**Comment by mesqueeb** [comment]
@bahmutov thanks for your guidance! 
What's the benefit of using fixtures over just importing mock objects directly into your test files ?

**Comment by sebastianjung** [comment]
@mesqueeb 
I think manipulating the imported data in one test will leave the data manipulated in other tests. That's why cloning is important.

> I could not get the task option to work, as my .js fixture files had ES6 imports that were constantly failing. In the end, I just did a standard import of each fixture file at the top of each spec file and then cloned the data in a beforeEach.

**Comment by sebastianjung** [comment]
> I could not get the task option to work, as my .js fixture files had ES6 imports that were constantly failing. In the end, I just did a standard import of each fixture file at the top of each spec file and then cloned the data in a beforeEach.
> 
> ```js
> import props from '../../../fixtures/Models/Helpers/props';
> import _ from 'lodash';
> 
> ....
> 
> 	beforeEach(function() {
> 		cy.then(function() {
> 			return _.cloneDeep(props);
> 		}).as('props');
> 	});
> ```
> 
> The fixture was then available in each test as **this.props**.

@OneHatRepo 
I am not able to access the data with this.props. Do i have to be in a specific scope or should i be able to use it within a it() closure?

Does the before each have to be inside the describe closure or is other advice you could possibly give?

Any help is very much appreciated!

**Comment by melibe23** [comment]
Hello! Any news about this bug? I will much appreciate the effort in a fix, our tests are getting bigger and bigger and are hard to maintain without this functionality.

**Comment by melibe23** [comment]
This is killing me. 
I need to use RegExp in a JSON due to this bug that Cypress have. Ok. But then, I found that match works with a literal RegExp but not with Object RegExp, and I don't know why.

**Comment by supasympa** [comment]
Hi, 

Is there any update on this?

I have a third party script that controls features within my application and I don't control how the external script works, it calls into the application and invokes a global method. The script is loaded by the main app and I'd like to `intercept` and replace the script in the tests. 

I'd like to be able to do it something like this: 

```Javascript
   cy.intercept('https://third-party.com/js/foo.js', {
      fixture: 'third-party-com//js/foo.js',
    }).as('3rd-party-foo')
...

cy.wait('@3rd-party-foo')
```

Is there an alternative way to achieve this?

**Comment by supasympa** [comment]
It looks like this works, not ideal but might help people out. 


```Javascript
cy.readFile('fixtures/3rd-party-scripts/3rdparty/js/foo.js', 'utf8' ).then((stubResponse) => {
    cy.intercept('https://3rdparty.com/js/foo.js', (req) => {
      req.reply(stubResponse)
    }).as('3rd-party-foo')
  })
  
  ...
  
    cy.wait('@3rd-party-foo')
  
```

**Comment by MohamadKh75** [comment]
> Maybe we need to update the document to state that js fixture is not what people normally expect.
> I am quite surprised that
> this fixture works:
> 
> ```js
> module.exports = {
>   a: 1
> }
> ```
> 
> but not this (with semicolon automatically added by my editor setting)
> 
> ```js
> module.exports = {
>   a: 1
> };
> ```
> 
> This throws `Error: 'myFixture' is not a valid JavaScript object.SyntaxError: Unexpected token ;`

Any update on this?
I need to add the `fixture` directory to `.prettierignore` 😐

**Comment by ahfarmer** [comment]
Seems like it would help a ton of people if there was a mechanism to load a fixture as-is without doing whatever special magic Cypress is attempting to do with `.js` files.

Like so:

```javascript
      cy.intercept("/myjs.js", {
        fixture: "stubjs.js",
        raw: true,
        headers: {
          "content-type": "application/javascript",
        },
      });
```

**Comment by poponuts** [comment]
any updates on this?

**Comment by MohamadKh75** [comment]
Hey
We really need to use js in our fixtures for such simple things (I. e. generate today date) or use [Faker](https://github.com/marak/Faker.js/)!

Any update on the estimated timeline or roadmap to fix this?
cc: @brian-mann @jennifer-shehane

Thanks in advance :heart:

**Comment by MohamadKh75** [comment]
> Thanks @bahmutov for mentioning `cy.task`.
>
> I just started with Cypress and I wanted to prevent the issue that @SLOBYYYY had. To do that, I created javascript files that exports an object and I get it with the following task:
>
> **plugins/index.js**
>
> ```js
> module.exports = (on, config) => {
>   on('task', {
>     fixture(fixtures) {
>       const result = {};
>
>       if (Array.isArray(fixtures)) {
>         fixtures.forEach(fixture => {
>           result[fixture] = require('../fixtures/elements/' +
>             fixture +
>             '.js');
>         });
>       } else {
>         result[fixtures] = require('../fixtures/elements/' +
>           fixtures +
>           '.js');
>       }
>
>       return result;
>     }
>   });
>
>   return config;
> };
> ```
>
> **So in your tests you can use it like this**
>
> ```js
> // Getting one fixture
> cy.task('fixture', 'user').then(f => {
>   console.log(f); // outputs {user: {}}
> });
>
> // Getting multiple fixtures
> cy.task('fixture', ['user', 'product']).then(f => {
>   console.log(f); // outputs { user: {}, product: {]}
> });
> ```
>
> This way you can also use libraries like faker in your `.js` files. You can adjust the task so that it fits your project.

There was some error when I tried this so here is my version (tested on Cypress@8.3.1):

**plugins/index.js**:

```js
module.exports = (on, config) => {
  on('task', {
    fixture({ name, address }) {
      const result = require(`../fixtures/${address}`);
      return { result, name };
    }
  });

  return config;
};
```

**How to use it**:

```js
cy.task('fixture', {
  name: 'someName',
  address: 'someWhere/file.js'
}).then(function({ result, name }) {
  this[name] = result;
});
```

**Finally**:

```js
cy.get('@something').should('contain.text', this.someName.x);
```

**Comment by aryankarim** [comment]
I wanna import a JSON file outside fixtures via a JavaScript code. I tried this requiring it into an array but it doesn't work.
e.g. 
// fixtures/data.js 
`[require('../../data.json')]`

is that even possible?

**Comment by hloughrey** [comment]
I was able to get factory.ts & faker working within a fixture with the following:


```ts
// fixtures/someFile.ts

const MyFactory = Factory.Sync.makeFactory<--- some TS type --->({
  id: faker.datatype.string(),
  name: faker.random.arrayElement<string>(--- some array ---),
  types: faker.random.arrayElements<string>(--- some array ---, 2),
});

export const myFixture: <--- some TS type ---> = {
  types: faker.random.arrayElements<string>(--- some array ---, 2),
  data: [
    MyFactory.build(),
    MyFactory.build(),
    MyFactory.build(),
    MyFactory.build(),
  ],
};
```

then in my spec file:

```ts
import { myFixture } from '../fixtures';

describe('Some Module', () => {
   beforeEach(() => {
       cy.intercept({ method: 'POST', path: '/some/api/endpoint' }, myFixture).as('someAlias');
   })
   ...
})
```

Hope this helps someone. 👍

**Comment by Maxim-Mazurok** [comment]
This is my solution:
```js
export const interceptFileJS = () => {
    global.myCypressCache = global.myCypressCache || {};
    const originalFile = Promise.resolve(global.myCypressCache.file) || cy.readFile('cypress/fixtures/file.js');
    originalFile.then((script) => {
        global.myCypressCache.file = script;
        script = script.replace(/%variable%/, 'something else');
        cy.intercept(`/assets/vendor/file.min.js`, { body: script }).as('fileJs');
    });
};
```
Primitive caching included, the downside is using global scope, but should be fine.

Alternative solution is to embed script in test:
```js
cy.intercept(`/assets/vendor/file.js`, {
    body: `alert('my script lives here')`,
}).as('fileJs');
```

The downside is no formatting/linting/etc.
